src/main/java/org/reactome/addlinks/fileprocessors/ensembl/EnsemblBioMartFileProcessor.java

```diff
@@ -0,0 +1,167 @@
+             |              |
+             |              +---ENSGALP00018
+             |                      |
+             |                      + ENSGALT00007, ENSGALT00008,...
+             +hsapiens_transcriptsToMicroarrays
+             |              |
+             |              +---ENST000019
+             |              |       |
+             |              |       + ILMN_123456, ILMN_7891011,...
+             |              |
+             |              +---ENST000020
+             |                      |
+             |                      + ILMN_121314, ILMN_141617,...
+             |
+             +mmusculus_proteinsToGenes
+             |              |
+             |              +---ENSMUSP000020
+             |              |       |
+             |              |       + ENSMUSG00009, ENSMUSG00010,...
+             |              |
+             |              +---ENSMUSP000021
+             |                      |
+             |                      + ENSMUSG00011, ENSMUSG00012,...
+
+         */
+
+        Map<String, Map<String, List<String>>> mappings = new HashMap<>();
+        // Read each file in BioMart directory and iterate through each line.
+        try (Stream<Path> paths = Files.walk(this.pathToFile)) {
+            paths
+                .filter(Files::isRegularFile)
+                .forEach(bioMartFile -> {
+                    // Get species name from filename
+                    String biomartFileSpecies = bioMartFile.getFileName().toString().split("_")[0];
+                    // Mappings that will be generated from each file for the species.
+                    Map<String, List<String>> proteinToTranscripts = new HashMap<>();
+                    Map<String, List<String>> proteinToGenes = new HashMap<>();
+                    Map<String, List<String>> transcriptToMicroarrays = new HashMap<>();
+                    Map<String, List<String>> uniprotToProteins = new HashMap<>();
+                    try {
+                        for (String biomartLine : Files.readAllLines(bioMartFile)) {
+                            // Split each tab-separated line. Each line has four values. The first three are always
+                            // Ensembl Gene (ENSG), Transcript (ENST) and Protein (ENSP), in that order.
+                            // The 4th can be a UniProt or microarray identifier, depending on the file being read.
+                            // None of the values are guaranteed to in the line, except perhaps for the first (Gene).
+                            List<String> tabSplit = Arrays.asList(biomartLine.split("\t"));
+                            // Processing of UniProt mapping files.
+                            // UniProt identifier mapping files are used to fully populate
+                            // the 'uniprotToProteins' and 'proteinToGenes' mappings.
+                            if (bioMartFile.toString().endsWith("uniprot")) {
```

Comment:
This check can be done outside the for loop iterating over the lines of the file since it depends only on the name of the file.  That way the check is only executed once.

Reply:
I'm not sure I understand what you mean. Immediately following the suffix check is when the pertinent data structures are populated. UniProt-ending files populate uniprotToProteins/proteinToGenes while Microarray-ending files populate proteinToTranscripts/TranscriptToMicroarrays. If I didn't separate these, then microarray/uniprot identifiers would corrupt each others data structures (since both UniProt and Microarray IDs are in the 4th column of their respective files). 

I could put the suffix check ahead of the line iteration, and then have a separate method (parseUniProtFile/parseMicroarrayFile). This would only add a small amount of code in its current state (since line iteration would take place in both of these new methods).